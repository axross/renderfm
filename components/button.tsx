import { twMerge } from "tailwind-merge";
import * as React from "react";
import { LoadingIcon } from "~/components/icon";
import { primaryFocusRing } from "~/constants/style";

export type ButtonVariant = "primary";
export type ButtonSize = "base" | "sm" | "xs";

const DEFAULT_VARIANT: ButtonVariant = "primary";
const DEFAULT_SIZE: ButtonSize = "base";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  translucent?: boolean;
  loading?: boolean;
}

export const Button: React.VFC<ButtonProps> = ({
  variant = DEFAULT_VARIANT,
  size = DEFAULT_SIZE,
  loading = false,
  translucent = false,
  disabled,
  className,
  children,
  ...props
}) => {
  return (
    <button
      disabled={disabled || loading}
      className={twMerge(
        "inline-flex items-center rounded-md disabled:cursor-not-allowed transition-all",
        variant === "primary"
          ? translucent
            ? "bg-yellow-600/20 hover:bg-yellow-600/30 disabled:bg-yellow-600/10 text-yellow-600"
            : "bg-yellow-600 disabled:bg-yellow-600/50 text-white"
          : null,
        variant === "primary" ? primaryFocusRing : null,
        size === "base" ? "px-4 py-2 text-base" : null,
        size === "sm" ? "px-3 py-1.5 text-base" : null,
        size === "xs" ? "px-2 py-1 text-sm" : null,
        className
      )}
      {...props}
    >
      {loading ? (
        <LoadingIcon
          className={twMerge(
            size === "base" ? "w-5 h-5 -ml-1.5 mr-2" : null,
            size === "sm" ? "w-4 h-4 -ml-1 mr-1.5" : null,
            size === "xs" ? "w-3 h-3 mr-1" : null
          )}
        />
      ) : null}

      {children}
    </button>
  );
};
