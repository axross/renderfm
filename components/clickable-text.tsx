import React from "react";
import { twMerge } from "tailwind-merge";
import { primaryFocusRing } from "~/constants/style";

const clickableTextCss = twMerge(
  primaryFocusRing,
  "rounded-sm text-yellow-600 dark:text-yellow-400 hover:underline"
);

export interface ClickableTextProps {
  onClick?: React.MouseEventHandler<HTMLSpanElement>;
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode;
}

export const ClickableText: React.VFC<ClickableTextProps> = ({
  className,
  children,
  ...props
}) => {
  return (
    <span
      className={twMerge(clickableTextCss, "cursor-pointer", className)}
      tabIndex={0}
      {...props}
    >
      {children}
    </span>
  );
};

export type TextAnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement>;

export const TextAnchor: React.VFC<TextAnchorProps> = ({
  href,
  className,
  children,
  ...props
}) => {
  return (
    <a href={href} className={twMerge(clickableTextCss, className)} {...props}>
      {children}
    </a>
  );
};
