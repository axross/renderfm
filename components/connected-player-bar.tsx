import { twMerge } from "tailwind-merge";
import { PlayerBar } from "~/components/player-bar";
import { useHotkeyEffect } from "~/hooks/use-hotkey-effect";
import { useAudioPlayer } from "use-audio-player";
import React from "react";

interface ConnectedPlayerBar {
  className?: string;
  style?: React.CSSProperties;
}

export const ConnectedPlayerBar: React.VFC<ConnectedPlayerBar> = ({
  className,
  ...props
}) => {
  const player = useAudioPlayer();

  useHotkeyEffect(
    "ArrowLeft",
    () => {
      if (!player.isPlayable) return;

      player.seek(-15);
    },
    [player]
  );

  useHotkeyEffect(
    "ArrowRight",
    () => {
      if (!player.isPlayable) return;

      player.seek(15);
    },
    [player]
  );

  useHotkeyEffect(
    " ",
    () => {
      if (!player.isPlayable) return;

      if (player.isPlaying) {
        player.pause();
      } else {
        player.play();
      }
    },
    [player]
  );

  return (
    <PlayerBar
      trackTitle={player.track?.title ?? ""}
      duration={player.duration}
      currentTime={player.currentTime}
      playing={player.isPlaying}
      durationLoaded={player.isDurationLoaded}
      playbackLoading={player.isPlaybackLoading}
      onPauseButtonClick={player.pause}
      onResumeButtonClick={player.play}
      onRewindButtonClick={React.useCallback(() => player.seek(-15), [player])}
      onForwardButtonClick={React.useCallback(() => player.seek(15), [player])}
      onSeek={(d) => player.seekTo(player.duration * d)}
      onCloseButtonClick={player.unload}
      className={twMerge(
        "fixed left-0 right-0 transition-all",
        player.track !== null
          ? "bottom-0"
          : "-bottom-40 pointer-fine:-bottom-24",
        className
      )}
      {...props}
    />
  );
};
