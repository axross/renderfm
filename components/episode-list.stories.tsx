import { ComponentStory, ComponentMeta } from "@storybook/react";
import { EpisodeList, EpisodeListRow } from "~/components/episode-list";

export default {
  title: "Components/EpisodeList",
  component: EpisodeList,
  subcomponents: { EpisodeListRow },
  args: {
    className: "w-96",
    children: (
      <>
        <EpisodeListRow
          episode={{
            publishedAt: "2022-01-01T00:00:00-0700",
            duration: 1000,
            title: "Lorem ipsum dolor sit amet",
            summaryPlainText:
              "Maecenas sed enim ut sem viverra aliquet eget sit amet. Mattis aliquam faucibus purus in massa tempor. Vitae ultricies leo integer malesuada.",
          }}
        />
        <EpisodeListRow
          episode={{
            publishedAt: "2021-01-01T00:00:00-0700",
            duration: 1000,
            title: "Ut enim",
            summaryPlainText:
              "ああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああああ",
          }}
        />
        <EpisodeListRow
          episode={{
            publishedAt: "2020-01-01T00:00:00-0700",
            duration: 1000,
            title:
              "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur",
            summaryPlainText: "Pellentesque.",
          }}
        />
      </>
    ),
  },
} as ComponentMeta<typeof EpisodeList>;

export const List: ComponentStory<typeof EpisodeList> = (args) => (
  <EpisodeList {...args} />
);

export const Row: ComponentStory<typeof EpisodeListRow> = (args) => (
  <EpisodeListRow {...args} />
);
Row.args = {
  episode: {
    publishedAt: "2022-01-01T00:00:00-0700",
    duration: 1000,
    title: "Lorem ipsum dolor sit amet",
    summaryPlainText:
      "Maecenas sed enim ut sem viverra aliquet eget sit amet. Mattis aliquam faucibus purus in massa tempor. Vitae ultricies leo integer malesuada.",
  },
};
