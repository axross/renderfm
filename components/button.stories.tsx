import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "~/components/button";

export default {
  title: "Components/Button",
  component: Button,
  args: {
    variant: "primary",
    disabled: false,
    children: "Press Me",
  },
} as ComponentMeta<typeof Button>;

export const Example: ComponentStory<typeof Button> = (args) => (
  <Button {...args} />
);

export const Small = Example.bind({});
Small.args = {
  size: "sm",
};

export const ExtraSmall = Example.bind({});
ExtraSmall.args = {
  size: "xs",
};

export const Translucent = Example.bind({});
Translucent.args = {
  translucent: true,
};

export const Disabled = Example.bind({});
Disabled.args = {
  disabled: true,
};

export const Loading = Example.bind({});
Loading.args = {
  loading: true,
};
