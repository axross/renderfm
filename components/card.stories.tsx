import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "~/components/button";
import { Card, CardFooter } from "~/components/card";

export default {
  title: "Components/Card",
  component: Card,
  args: {
    className: "w-96 min-h-64",
    children: "Hello",
  },
} as ComponentMeta<typeof Card>;

export const Example: ComponentStory<typeof Card> = (args) => (
  <Card {...args} />
);

export const WithFooter = Example.bind({});
WithFooter.args = {
  children: (
    <>
      Hello
      <CardFooter>
        <Button>Do Some Action</Button>
      </CardFooter>
    </>
  ),
};
