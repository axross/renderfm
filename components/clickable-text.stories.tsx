import { ComponentStory, ComponentMeta } from "@storybook/react";
import { ClickableText } from "~/components/clickable-text";

export default {
  title: "Components/ClickableText",
  component: ClickableText,
  args: {
    children: "This is a link",
  },
} as ComponentMeta<typeof ClickableText>;

export const Example: ComponentStory<typeof ClickableText> = (args) => (
  <ClickableText {...args} />
);
