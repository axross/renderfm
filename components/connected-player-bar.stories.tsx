import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "~/components/button";
import { ConnectedPlayerBar } from "~/components/connected-player-bar";
import {
  AudioPlayerProvider,
  createBrowserAudioPlayer,
  useAudioPlayer,
} from "use-audio-player";

export default {
  title: "Connected Components/PlayerBar",
  component: ConnectedPlayerBar,
  decorators: [
    (Story) => {
      const player = createBrowserAudioPlayer();

      return (
        <AudioPlayerProvider player={player}>
          <Story />
        </AudioPlayerProvider>
      );
    },
  ],
} as ComponentMeta<typeof ConnectedPlayerBar>;

export const Example: ComponentStory<typeof ConnectedPlayerBar> = (args) => {
  const player = useAudioPlayer();

  return (
    <>
      <Button
        onClick={() => {
          player.load({
            title: "Some Fancy Music",
            src: "https://file-examples-com.github.io/uploads/2017/11/file_example_MP3_5MG.mp3",
          });

          // eslint-disable-next-line storybook/await-interactions, storybook/context-in-play-function
          player.play();
        }}
      >
        Play Something
      </Button>

      <ConnectedPlayerBar {...args} />
    </>
  );
};
