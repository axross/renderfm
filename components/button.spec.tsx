import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Button, ButtonProps } from "~/components/button";

describe("<Button>", () => {
  function setup(additionalProps?: Partial<ButtonProps>) {
    const onClick = jest.fn();

    render(
      <Button onClick={onClick} data-testid="button" {...additionalProps}>
        Click me
      </Button>
    );

    return {
      button: () => screen.getByTestId("button"),
      onClick,
    };
  }

  const setupDisabled: typeof setup = (additionalProps) =>
    setup({ disabled: true, ...additionalProps });

  const setupLoading: typeof setup = (additionalProps) =>
    setup({ loading: true, ...additionalProps });

  it("renders a medium button by default", () => {
    const { button } = setup();

    expect(button()).toMatchSnapshot();
  });

  it('renders a small button when size="sm" is given', () => {
    const { button } = setup({ size: "sm" });

    expect(button()).toMatchSnapshot();
  });

  it('renders an extra small button when size="xs" is given', () => {
    const { button } = setup({ size: "xs" });

    expect(button()).toMatchSnapshot();
  });

  it("renders a translucent button when transulucent=true is given", () => {
    const { button } = setup({ translucent: true });

    expect(button()).toMatchSnapshot();
  });

  it("renders a button focusable", () => {
    const { button } = setup();

    userEvent.tab();

    expect(button()).toHaveFocus();

    userEvent.tab();

    expect(button()).not.toHaveFocus();
  });

  it("renders a button unfocusable when disabled=true is given", () => {
    const { button } = setupDisabled();

    userEvent.tab();

    expect(button()).not.toHaveFocus();
  });

  it("calls onClick() when the user clicks it", () => {
    const { button, onClick } = setup();

    userEvent.click(button());

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("doesn't call onClick() even when the user clicks it as long as disabled=true is given", () => {
    const { button, onClick } = setupDisabled();

    userEvent.click(button());

    expect(onClick).not.toHaveBeenCalled();
  });

  it("renders a button with the children alongwith a loding spinner when loading=true is given", () => {
    const { button } = setupLoading();

    expect(button()).toMatchSnapshot();
  });

  it("renders a button unfocusable when loading=true is given", () => {
    const { button } = setupLoading();

    userEvent.tab();

    expect(button()).not.toHaveFocus();
  });

  it("doesn't call onClick() even when it gets clicked as long as loading=true is given", () => {
    const { button, onClick } = setupLoading();

    userEvent.click(button());

    expect(onClick).not.toHaveBeenCalled();
  });
});
