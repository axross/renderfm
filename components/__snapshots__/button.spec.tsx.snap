// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> renders a button with the children alongwith a loding spinner when loading=true is given 1`] = `
<button
  class="inline-flex items-center rounded-md disabled:cursor-not-allowed bg-yellow-600 disabled:bg-yellow-600/50 text-white outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all px-4 py-2 text-base"
  data-testid="button"
  disabled=""
>
  <svg
    class="animate-spin w-5 h-5 -ml-1.5 mr-2"
    fill="none"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="opacity-25"
      cx="12"
      cy="12"
      r="10"
      stroke="currentColor"
      stroke-width="4"
    />
    <path
      class="opacity-75"
      d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
      fill="currentColor"
    />
  </svg>
  Click me
</button>
`;

exports[`<Button> renders a medium button by default 1`] = `
<button
  class="inline-flex items-center rounded-md disabled:cursor-not-allowed bg-yellow-600 disabled:bg-yellow-600/50 text-white outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all px-4 py-2 text-base"
  data-testid="button"
>
  Click me
</button>
`;

exports[`<Button> renders a small button when size="sm" is given 1`] = `
<button
  class="inline-flex items-center rounded-md disabled:cursor-not-allowed bg-yellow-600 disabled:bg-yellow-600/50 text-white outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all px-3 py-1.5 text-base"
  data-testid="button"
>
  Click me
</button>
`;

exports[`<Button> renders a translucent button when transulucent=true is given 1`] = `
<button
  class="inline-flex items-center rounded-md disabled:cursor-not-allowed bg-yellow-600/20 hover:bg-yellow-600/30 disabled:bg-yellow-600/10 text-yellow-600 outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all px-4 py-2 text-base"
  data-testid="button"
>
  Click me
</button>
`;

exports[`<Button> renders an extra small button when size="xs" is given 1`] = `
<button
  class="inline-flex items-center rounded-md disabled:cursor-not-allowed bg-yellow-600 disabled:bg-yellow-600/50 text-white outline-none focus:ring-4 focus:ring-yellow-500/50 transition-all px-2 py-1 text-sm"
  data-testid="button"
>
  Click me
</button>
`;
